sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -858,6 +859,30 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+  * Push [[Limit]] operators through [[Join]] operators, iff the join type is outer joins.
+  * Adding extra [[Limit]] operators on top of the outer-side child/children.
+  */
+object PushLimitThroughOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
For the `full outer` join, my idea is to push down one side which has a higher `statistics`. Does that sound good?